keps/sig-api-machinery/35-storage-version-hash.md

```diff
@@ -0,0 +1,108 @@
+---
+kep-number: 35
+title: exposing hashed storage versions via the discovery API
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2010-01-02
+last-updated: 2018-xx-xx
+status: provisional
+---
+
+# Exposing storage versions in opaque values via the discovery API
+
+## Table of Contents
+
+## Terms
+
+**storage versions**: Kubernetes API objects are converted to specific API
+versions before stored in etcd. "Storage versions" refers to these API versions.
+
+## Summary
+
+We propose to expose the hashed storage versions in the discovery API.
+
+## Motivation
+
+We intend to use the exposed storage version hash to trigger the [storage
+version migrator][].
+
+In short, the storage version migrator detects if objects in etcd are stored in
+a version different than the configured storage version. If so, the migrator
+issues no-op update for the objects to migrate them to the storage version.
+
+The storage version migrator can keep track of the versions the objects are
+stored as. However, today the migrator has no way to tell what the expected
+storage versions are. Thus we propose to expose this piece of information via
+the discovery API.
+
+[storage version migrator]:https://github.com/kubernetes-sigs/kube-storage-version-migrator
+
+## Proposal
+
+### API changes to the discovery API
+
+We add a new field `StorageVersionHash` to the [APIResource][] type.
+
+We use the hash value instead of the plain text to avoid clients misusing the
+information. So far, the only valid consumer of this information is the storage
+version migrator, which only needs to do equality comparisons on the storage
+versions.
```

Comment:
what are the desired properties of the StorageVersionHash? like it only needs to be unique per resource.

Reply:
It needs to be unique for different storage versions. SHA-256 is good enough. Added to the KEP.